### ISSUE TITLE
Handle (custom) Type parsing errors better

### ISSUE THREAD

**Issue by zsoldosp** [behave/behave#383]
## Problem summary
```python
from behave import register_type
def my_custom_type(text):
    raise Exception('could not parse text')
register_type(Thing=my_custom_type)
```
if we have the above custom type registered, when running a set of feature files, as soon as this type is attempted to be parsed for a step, the whole `behave` process crashes and no report is printed. 

I think it shouldn't crash the process.

Below is a feature file reproducing the issue, as well as a sketch of a potential solution (which isn't yet a full solution) - I'm happy to work on this to develop this into a pull request with some guidance from a core developer

----
## Repro feature file

`issue.features/issueXXXX.feature`
```
@issue
Feature: Issue #XXX: Exception from a step parser can crash the whole behave process

  Scenario: Does not crash the process
    Given a new working directory
    And   a file named "features/environment.py" with:
      """
      from behave import register_type

      def my_custom_type(text):
          raise Exception('could not parse text')

      register_type(Thing=my_custom_type)
      """
    And   a file named "features/steps/steps.py" with:
      """
      @when(u'a custom thing {something:Thing} is created')
      def step_impl(context, something):
          pass

      @then(u'all is OK')
      def step_impl(context):
          pass
      """
    And   a file named "features/bad-parser-shouldnt-crash-process.feature" with:
      """
      Feature:
        Scenario: after parsing fails in when, then step should be executed
          When  a custom thing foo is created
          Then  all is OK
      """
    When I run "behave -f plain "features/bad-parser-shouldnt-crash-process.feature""
    Then it should fail with:
      """
      0 scenarios passed, 1 failed, 0 skipped
      0 steps passed, 1 failed, 1 skipped, 0 undefined
      """
    And the command output should contain:
        """
           Exception('could not parse text')
        """
```

## A possible solution idea

```diff
diff --git a/behave/matchers.py b/behave/matchers.py
index 22beee8..868e633 100644
--- a/behave/matchers.py
+++ b/behave/matchers.py
@@ -3,6 +3,7 @@ from __future__ import absolute_import, with_statement
 import re
 import parse
 from parse_type import cfparse
+import six
 from behave import model


@@ -72,7 +73,11 @@ class ParseMatcher(Matcher):
         self.parser = parse.compile(self.string, self.custom_types)

     def check_match(self, step):
-        result = self.parser.parse(step)
+        try:
+            result = self.parser.parse(step)
+        except CustomTypeParsingFailed:
+            # TODO: need to print exception to stdout somehow
+            return []
         if not result:
             return None

@@ -96,6 +101,9 @@ class CFParseMatcher(ParseMatcher):
         super(ParseMatcher, self).__init__(func, string, step_type)
         self.parser = cfparse.Parser(self.string, self.custom_types)

+class CustomTypeParsingFailed(Exception):
+    pass
+

 def register_type(**kw):
     """Registers a custom type that will be available to "parse"
@@ -127,7 +135,18 @@ def register_type(**kw):
         def step_impl(context, amount):
             assert isinstance(amount, int)
     """
-    ParseMatcher.custom_types.update(kw)
+    def custom_type_converter(orig_converter, text):
+        try:
+            return orig_converter(text)
+        except:
+            # TODO: pass in the original exception (traceback) into the exception
+            raise CustomTypeParsingFailed()
+
+    wrapped_converters = dict(
+        (type_name, lambda text: custom_type_converter(type_converter, text))
+        for (type_name, type_converter) in six.iteritems(kw)
+    )
+
+    ParseMatcher.custom_types.update(wrapped_converters)
```

**Comment by jenisys** [comment]
I will look a your solution, but my gut feeling is that the exception should be raised.
When a type parsing fails (with an exception), the feature file contains invalid data (type value) in a step. Therefore, the step should fail (and you should be able to understand what went wrong).

After using your example feature, I agree that this situation (errors/exceptions during type parsing) should be improved, but probably in the ``Step.run()`` method. Because there you can better provide the step context as information for the problem location.

**Comment by zsoldosp** [comment]
@jenisys yes, I had no intention of hiding these exceptions - see the `TODO`s I've left in there

I should have been more explicit - I wanted to know whether the custom exception raising was OK before proceeding further

I will look at `Step.run()`.

The initial limitation I've run into was that when the `check_match` method raises an exception, then behave interprets it as if no matching step could be found, and suggests the `step_impl` snippet on the stdout. Thus my "choice" in the proof of concept code to return an empty args list (so the code at least passes)

**Comment by jenisys** [comment]
FIX INTEGRATED INTO: behave-1.2.6.dev0
